### PR TITLE
node: Prepare development binary release

### PR DESCRIPTION
**Pull request by kcsongor**
In preparation for a development release of the node image, this PR introduces several changes:

* remove the cruft from the docker image by copying only the necessary files into a scratch stage
* add a `GO_BUILD_ARGS` build arg to the docker container which by default sets the `-race` flag, but allows turning it off
* enforce that `--unsafeDevMode` is enabled when running a development build of the binary.

---
This change is [Reviewable](https://reviewable.io/reviews/certusone/wormhole/1332)

### CODE REVIEW

File: node/Dockerfile [hunk 1]
```diff
@@ -21,4 +20,18 @@ RUN --mount=type=cache,target=/root/.cache --mount=type=cache,target=/go \
   go build -race -gcflags="all=-N -l" --ldflags '-extldflags "-Wl,--allow-multiple-definition"' -mod=readonly -o /guardiand github.com/certusone/wormhole/node && \
```
Comment (leoluk) [inline node/Dockerfile:20]: We don't want to run with `-race` in production

File: node/Dockerfile [hunk 2]
```diff
@@ -21,4 +20,18 @@ RUN --mount=type=cache,target=/root/.cache --mount=type=cache,target=/go \
...
   cp /go/pkg/mod/github.com/!cosm!wasm/wasmvm@v0.16.2/api/libwasmvm.so /usr/lib/
 
+# Only export the final binary (+ shared objects). This reduces the image size
+# from ~1GB to ~150MB.
+FROM scratch as export
+
+# guardiand can't (easily) be statically linked due to the C dependencies, so we
+# have to copy all the dynamic libraries
+COPY --from=build /bin/* /bin/
```
Comment (leoluk) [inline node/Dockerfile:29]: We'll want `/dlv` too if this same image should also run in tilt. Or specify that we want the `build` stage when building in Tilt

**Review by leoluk — CHANGES_REQUESTED** [review]
Perhaps a separate Dockerfile (or at least separate stages) would be easier?

File: node/Dockerfile (hunk 2 above)
Comment (kcsongor) [reply node/Dockerfile:29]: I'll specify that tilt should run `build`

File: node/Dockerfile (hunk 1 above)
Comment (kcsongor) [reply node/Dockerfile:20]: Do you have a preferred solution for doing this? I'm thinking a build arg like `NO_SANITIZER` that's set to false by default

**Comment by kcsongor** [comment]
> Perhaps a separate Dockerfile (or at least separate stages) would be easier?

I've kept the single docker file but made it a bit more configurable to support building without `-race`. Also added a way for the binary to check if it was built in development mode, and require the `--unsafeDevMode` flag in that case. Let me know what you think of doing it this way

**Review by SEJeff — APPROVED** [review]
Strong +1, nice!

**Review by SEJeff — APPROVED** [review]

File: .github/workflows/guardiand-docker.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Log in to the Container registry
+        uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
Comment (claudijd) [inline .github/workflows/guardiand-docker.yml:28]: Should we use the default GITHUB_TOKEN permissions or generate a special key for this action and scope down it's permissions to a minimal set of scope?

Comment (kcsongor) [reply .github/workflows/guardiand-docker.yml:28]: No need for that, I've specified in the workflow file that the token should only have write access to the packages (where it publishes): https://github.com/certusone/wormhole/pull/1332/files#diff-0646c51b93a786d348c46b78d911a1c9d066b8263e4f0dfc817839bdf86fd0c1R15-R17

reference: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token
> You can use the `permissions` key in your workflow file to modify permissions for the GITHUB_TOKEN for an entire workflow or for individual jobs. This allows you to configure the minimum required permissions for a workflow or job. When the `permissions` key is used, all unspecified permissions are set to no access, with the exception of the metadata scope, which always gets read access.

**Review by leoluk — APPROVED** [review]

**Pull request merged**